hathor/websocket/streamer.py

```diff
@@ -0,0 +1,196 @@
+            idx += 1
+
+
+class GAPLimitSearch(AsyncIterable[tuple[str, Any]]):
+    def __init__(self, manager: HathorManager, address_iter: AsyncIterable[tuple[int, AddressB58]], gap_limit: int):
+        self.manager = manager
+        self.address_iter = address_iter
+        self.gap_limit = gap_limit
+
+    def __aiter__(self) -> AsyncIterator[tuple[str, Any]]:
+        return self._async_iter()
+
+    async def _async_iter(self) -> AsyncIterator[tuple[str, Any]]:
+        assert self.manager.tx_storage.indexes is not None
+        assert self.manager.tx_storage.indexes.addresses is not None
+        addresses_index = self.manager.tx_storage.indexes.addresses
+        empty_addresses_counter = 0
+        async for address_idx, address in self.address_iter:
+            yield ('address', (address_idx, address))
+
+            vertex_counter = 0
+            for vertex_id in addresses_index.get_sorted_from_address(address):
+                tx = self.manager.tx_storage.get_transaction(vertex_id)
+                yield ('vertex', tx.to_json_extended())
+                vertex_counter += 1
+
+            if vertex_counter == 0:
+                empty_addresses_counter += 1
+                if empty_addresses_counter >= self.gap_limit:
+                    break
+            else:
+                empty_addresses_counter = 0
+
+
+@implementer(IPushProducer)
+class HistoryStreamer:
+    def __init__(self,
+                 *,
+                 protocol: 'HathorAdminWebsocketProtocol',
+                 stream_id: str,
+                 search: GAPLimitSearch) -> None:
+        self.protocol = protocol
+        self.stream_id = stream_id
+        self.search_iter = aiter(search)
+
+        self._paused = False
+        self._stop = False
+
+    def start(self) -> Deferred[bool]:
+        self.send_message({'type': 'stream:history:begin', 'id': self.stream_id})
+
+        # The websocket connection somehow instantiates an twisted.web.http.HTTPChannel object
+        # which register a producer. It seems the HTTPChannel is not used anymore after switching
+        # to websocket but it keep registered. So we have to unregister before registering  a new
+        # producer.
+        self.protocol.unregisterProducer()
+
+        self.protocol.registerProducer(self, True)
+        self.deferred: Deferred[bool] = Deferred()
+        self.resumeProducing()
+        return self.deferred
+
+    def stop(self, success: bool) -> None:
+        self.protocol.unregisterProducer()
+        self.deferred.callback(success)
+
+    def pauseProducing(self) -> None:
+        self._paused = True
+
+    def stopProducing(self) -> None:
+        self._stop = True
+        self.stop(False)
+
+    def resumeProducing(self) -> None:
+        self._paused = False
+        self._run()
+
+    def _run(self) -> None:
+        coro = self._async_run()
+        Deferred.fromCoroutine(coro)
+
+    async def _async_run(self):
+        async for _type, data in self.search_iter:
```

Comment:
Should we limit the number of iterations to get back to the main loop? The current implementation waits for a call to `pauseProducing()` which is actually called by the transport `write()` method when the internal buffer gets full. The main loop might be locked for a long time if the buffer never gets full (e.g., very high bandwidth or very slow disk reading).

Reply:
I added some code that allows the event loop to run at least once every `N` seconds. What would be a good value for `N`? 1 second?

Reply:
I ran some tests and it seems the producer gets paused many times when running the full node and the client in the same network. I'd say 1 second is safe. But I've also tested with 0.1 seconds and it had a small impact (+10%) when loading an address with 500k+ transactions.

Reply:
Ok. Thanks! It's been working fine with 1 second. We can change it later if needed.